File: apps/server-core/src/room/target-caps.ts
// SPEC-REF:
//   docs/strategy/2026-09-06-web-client-parity-privacy-image-and-ux-addendum.md §3
//   packages/protocol/src/protocol-schemas-auth.ts (TargetCapsSchema — the shape)
//   card S2-01
//
// The ONE place `pc_devices.target_caps` crosses between a stored string and a
// wire object, in both directions.
//
// 🔴 WHY A MODULE AND NOT TWO `JSON.parse` CALLS. This column has THREE states —
// undeclared / declared-yes / declared-no — and only one of them is spelled by
// the JSON. The other two are「the column is NULL」and「the column holds
// something this build cannot read」, and both of those have to arrive at the
// SAME answer (undeclared) or the microphone end starts refusing images to
// targets that never said no. A `JSON.parse` at each call site is two chances to
// answer that question differently, and a `try/catch` at only one of them is the
// bug shipping.
//
// ⚠️ IT NEVER RETURNS `{image:false}` AS A FALLBACK. A stored value we cannot
// parse is not a target that declined — it is a target we have not heard from,
// and this repo's rule for that is to degrade to today's product (images flow),
// never to a new refusal nobody can act on.

import { TargetCapsSchema, type TargetCaps } from '@flowmic/protocol';

/**
 * Wire object → the string that goes in the column. `undefined` in means
 * `null` out: a client that declared nothing must leave the column exactly as a
 * pre-S2-01 row has it, so the two are indistinguishable downstream — which is
 * the whole meaning of「undeclared」.
 */
export function serializeTargetCaps(caps: TargetCaps | undefined): string | null {
  return caps === undefined ? null : JSON.stringify(caps);
}

/**
 * Column → the wire object, or `undefined` for「this target has not declared」.
 *
 * Three inputs collapse to `undefined` and that is deliberate: NULL (never
 * declared), unparseable text (a write from some future or broken build), and a
 * shape this build's schema rejects. All three mean the same thing to a reader —
 * we do not know — and inventing a different answer for any of them would be
 * asserting something nobody said.
 */
export function parseTargetCaps(stored: string | null): TargetCaps | undefined {
  if (stored === null || stored === '') return undefined;
  let raw: unknown;
  try {
    raw = JSON.parse(stored);
  } catch {
    return undefined;
  }
  const parsed = TargetCapsSchema.safeParse(raw);
  return parsed.success ? parsed.data : undefined;
}

File: apps/mobile/lib/src/signaling/target_caps.dart
// SPEC-REF:
//   packages/protocol/src/protocol-schemas-auth.ts (TargetCapsSchema,
//     TargetCapsAckFieldsSchema — the wire shape)
//   docs/strategy/2026-09-06-web-client-parity-privacy-image-and-ux-addendum.md §3
//   card S2-01
//
// What the TARGET at the other end of this room can receive, read off the
// `mobile:pair` / `mobile:reconnect` ack.
//
// 🔴 THREE STATES, AND THE THIRD IS THE COMMON ONE.
//   · undeclared — the ack carried no `target_caps`. Every FLOWMIC-PC built
//     before this card is here, and so is every relay that predates the field.
//     The rule is ALLOW: refusing would stop image delivery for every install on
//     the day this ships.
//   · declared yes — `{image: true}`.
//   · declared no — `{image: false}`. The microphone end says so BEFORE sending
//     and does not queue the picture (design §3: 「不入队、不计配额」).
// Collapsing undeclared into "no" is the tempting simplification and it is a
// regression; collapsing it into "yes" loses the ability to ever say no.
//
// 🔴 A DIFFERENT QUESTION FROM `ServerCapabilities`, which rides the same acks.
// That one says what the SERVER can do; this says what the TARGET can receive.
// The two even fail in opposite directions — an unknown server capability is
// read fail-CLOSED (hold the audio), an unknown target capability fail-OPEN
// (send the picture) — so one type carrying both would be a value answering two
// questions with two right answers.
//
// ⚠️ STORED, NOT YET ACTED ON. Nothing reads [canSendImage] in production yet:
// the image UI that asks before sending is card S3-02. That is deliberate and it
// is why this file ships with the reader — a capability the phone cannot see is
// a capability the phone cannot start honouring.

import 'package:flutter/foundation.dart' show immutable;

/// What one ack said about the target.
@immutable
class TargetCaps {
  const TargetCaps._(this.image, this.imageNote, this.declared);

  /// No ack read yet, or an ack that carried no `target_caps` key.
  const TargetCaps.undeclared()
      : image = null,
        imageNote = null,
        declared = false;

  /// Can the target receive a picture? `null` means it has not said.
  final bool? image;

  /// The target's OWN words for why not (a third-party site's setting), shown
  /// beside our sentence. Null unless the target supplied one — we never invent
  /// a reason on its behalf.
  final String? imageNote;

  /// Did the ack carry the key at all? See the file header for why this is not
  /// the same question as [image] being false.
  final bool declared;

  /// The rule the sending path will use (card S3-02): send unless the target
  /// has DECLARED it cannot. Undeclared allows, because that is where every
  /// installed target sits and because degrading to today's product is the only
  /// safe direction for a field a relay may strip.
  bool get canSendImage => image != false;

  @override
  String toString() => declared
      ? 'TargetCaps(image: $image${imageNote == null ? '' : ', note: $imageNote'})'
      : 'TargetCaps(undeclared)';
}

/// Read `target_caps` off a pair / reconnect ack.
///
/// TOLERANT IN ONE DIRECTION ONLY, the same way `parseServerCapabilities` is: a
/// missing key, a non-map, or a `image` that is not a real boolean all come back
/// as [TargetCaps.undeclared] rather than throwing — a malformed capability must
/// never break a pairing that is otherwise fine, and it must never be mistaken
/// for a refusal. What this will NOT do is invent a declaration: `image` has to
/// arrive as a genuine `true`/`false` to be one.
TargetCaps parseTargetCaps(Object? ack) {
  if (ack is! Map) return const TargetCaps.undeclared();
  final Object? raw = ack['target_caps'];
  if (raw is! Map) return const TargetCaps.undeclared();
  final Object? image = raw['image'];
  if (image is! bool) return const TargetCaps.undeclared();
  // TRIMMED, and a note that is only whitespace is dropped. The wire schema
  // accepts it (`NonEmpty` counts characters, and a space is one), so this is
  // the side that has to notice — a blank reason rendered beside our sentence
  // looks like a target that explained itself and said nothing.
  final Object? note = raw['image_note'];
  final String trimmed = note is String ? note.trim() : '';
  return TargetCaps._(image, trimmed.isEmpty ? null : trimmed, true);
}
